feat:  add  `math/base/special/atan2f`

Neerajpathak07 opened the pull request:
Progresses #649 

## Description

> What is the purpose of this pull request?

This pull request:

-   proposes adding `math/base/special/atan2f` , which would be a single precision implementation for [math/base/special/atan2](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/atan2).

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #649 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-  [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Neerajpathak07 committed:
feat: basic files added

Neerajpathak07 committed:
feat: basic c files added

Neerajpathak07 committed:
feat: test added

Neerajpathak07 committed:
feat: docs added

Neerajpathak07 committed:
feat: examples added

Neerajpathak07 committed:
feat: benchmark  added

Neerajpathak07 commented:
This is just a draft PR I am still working on it!!

Neerajpathak07 committed:
feat: lib added

Neerajpathak07 committed:
fix: test errors

Neerajpathak07 committed:
fix: lint coyright years

aayush0325 commented:
i see that this PR is failing a lot of tests. you should update your `manifest.json` to include all the necessary single-precsion dependencies to use them in `src/` and `lib/` currently you are including double precision dependencies but using the single-precision ones which is why this is causing an error.

aayush0325 commented on lib/node_modules/@stdlib/math/base/special/atan2f/manifest.json:
```diff
@@ -0,0 +1,91 @@
+        "libraries": [],
+        "libpath": [],
+        "dependencies": [
+          "@stdlib/math/base/napi/binary",
```
```suggestion
          "@stdlib/math/base/assert/is-infinitef",
          "@stdlib/math/base/special/copysignf",
          "@stdlib/number/float32/base/signbit",
          "@stdlib/math/base/assert/is-nanf",
          "@stdlib/math/base/special/atan2f",
          "@stdlib/constants/float32/pinf",
          "@stdlib/constants/float32/pi"

aayush0325 replied:
```suggestion
"@stdlib/math/base/assert/is-infinitef",
"@stdlib/math/base/special/copysignf",
"@stdlib/number/float32/base/signbit",
"@stdlib/math/base/assert/is-nanf",
"@stdlib/math/base/special/atan2f",
"@stdlib/constants/float32/pinf",
"@stdlib/constants/float32/pi"
```

aayush0325 commented:
```json
"dependencies": [
        "@stdlib/math/base/napi/binary",
        "@stdlib/math/base/assert/is-infinitef",
        "@stdlib/math/base/special/copysignf",
        "@stdlib/number/float32/base/signbit",
        "@stdlib/math/base/assert/is-nanf",
        "@stdlib/math/base/special/atanf",
        "@stdlib/constants/float32/pinf",
        "@stdlib/constants/float32/pi"
      ]
```


this is how you should fill the dependencies in `manifest.json`

Neerajpathak07 commented:
@aayush0325 thanks for the assist.

Neerajpathak07 committed:
fix: update manifest.json

aayush0325 commented:
use `number/float32/base/signbit` instead of `number/float32/base/signbitf` as the latter doesn't exist, checkout the package [here](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/number/float32/base/signbit)

Neerajpathak07 committed:
fix: benchmark errors

Neerajpathak07 commented:
@aayush0325 yeah I understood where I was wrong.
 `number/float32/base/signbit` I was supposed to declare this package and call `signbitf` as it is the single-precision version of `signbit`. Also I was writing the tests in double-precision so I'll convert them into single-precision.

Neerajpathak07 committed:
fix: addressing CI failures

Planeshifter commented:
@Neerajpathak07 The issue here is that `@stdlib/number/float32/base/signbit` doesn't yet have a C implementation. This PR will have to wait until we have that C implementation.

Neerajpathak07 commented:
@Planeshifter The PR related to `float32/signbit` has been merged could you unblock this PR

Neerajpathak07 committed:
Merge branch 'develop' of https://github.com/Neerajpathak07/stdlib into add-atan2f

Neerajpathak07 committed:
Merge branch 'stdlib-js:develop' into add-atan2f

Neerajpathak07 committed:
chore: clean up

Neerajpathak07 committed:
chore: update main.c

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/atan2f/lib/main.js:
```diff
@@ -0,0 +1,163 @@
+*
+* ## Notice
+*
+* The original code, copyright and license are from [Go]{@link https://golang.org/src/math/atan2.go}. The implementation follows the original, but has been modified for JavaScript.
```
@kgryte 
Apparently there was no golang source link for `atan2f` which is why I have kept it `atan2` for now.

Neerajpathak07 committed:
fix: example error

Neerajpathak07 committed:
fix: example error

Neerajpathak07 committed:
chore: update main.c

Neerajpathak07 committed:
fix: revert previous chnages

Neerajpathak07 committed:
chore: update main.c

Neerajpathak07 committed:
fix: native test error

aayush0325 commented on lib/node_modules/@stdlib/math/base/special/atan2f/README.md:
````diff
@@ -0,0 +1,209 @@
+Computes the angle in the plane (in radians) between the positive x-axis and the ray from `(0,0)` to the point `(x,y)` for single-precision floating-point number.
+
+```c
+float out = stdlib_base_atan2f( 2.0,f 2.0f );
````
```suggestion
float out = stdlib_base_atan2f( 2.0f, 2.0f );
```

aayush0325 commented on lib/node_modules/@stdlib/math/base/special/atan2f/src/main.c:
```diff
@@ -0,0 +1,124 @@
+#include "stdlib/constants/float32/pinf.h"
+#include "stdlib/constants/float32/ninf.h"
+#include "stdlib/constants/float32/pi.h"
+#include <stdint.h>
```
```suggestion
```
i don't think this header is being used

aayush0325 commented on lib/node_modules/@stdlib/math/base/special/atan2f/test/fixtures/julia/negative_negative.json:
ahh, these fixtures are probably the cause of failing tests here, you've copied these over from the double precision implementation as I compared the both and there is no diff. You are bound to face errors if you test your single precision implementation against a double precision dataset. CD into the dir and generate these fixtures by running `runner.jl` you may need to setup Julia and install JSON package for that first

aayush0325 replied:
holding out on further review before this is resolved, this should be it i think

aayush0325 reviewed (commented):
basically you need to generate the fixtures using the julia script and not copy them over from the double precision implementation

aayush0325 commented:
The reason for the failing benchmarks is probably because you copied over the Makefile wrongly, use the one in `math/base/special/atan2/benchmark/c/native/`

aayush0325 commented:
[Makefile that you should be using](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/atan2/benchmark/c/native/Makefile). This adds the include dirs so that you don't run into the error that we're currently facing here

Neerajpathak07 committed:
chore: update makefile

kgryte commented on lib/node_modules/@stdlib/math/base/special/atan2f/benchmark/c/native/Makefile:
```diff
@@ -1,4 +1,4 @@
-#/
+
```
This change is not correct. The comment should begin with `#/`.

Neerajpathak07 committed:
chore: update bench

Neerajpathak07 committed:
chore: update fixtures

Neerajpathak07 commented on lib/node_modules/@stdlib/math/base/special/atan2f/test/fixtures/julia/runner.jl:
````diff
@@ -0,0 +1,82 @@
+```
+"""
+function gen( x, y, name )
+	z = Array{Float32}( undef, length(x) );
+	for i in eachindex(x)
+		z[ i ] = atan.( y[i], x[i] );
````
Julia doesn't have support for computing `atan2f` and `atan2` so have kept it `atan` cuz that's the only one we have. But have generated it for `float32`

Neerajpathak07 commented:
@aayush0325 Well the error still persists even after generating fixtures for julia.

Neerajpathak07 committed:
chore: update benchmark

Neerajpathak07 committed:
chore: update benchmark

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

Neerajpathak07 committed:
Merge branch 'add-atan2f' of https://github.com/Neerajpathak07/stdlib into add-atan2f

Neerajpathak07 committed:
chore: update fixtures

Neerajpathak07 committed:
chore: minor changes

Neerajpathak07 commented:
@Planeshifter Only two test cases are failing in `test.native.js` whilst every tests and running-clean in `test.js`. Is there something I am missing out??

Neerajpathak07 committed:
chore: lint module name

anandkaranubc commented:
Completed via https://github.com/stdlib-js/stdlib/pull/7081
